apps/studio/state/storage-explorer.tsx

```diff
@@ -1417,6 +1532,406 @@ function createStorageExplorerState({
+
+          try {
+            await moveStorageObject({
+              projectRef: state.projectRef,
+              bucketId: state.selectedBucket.id,
+              from: fromPath,
+              to: toPath,
+            })
+          } catch (error: any) {
+            numberOfFilesMovedFail += 1
+            toast.error(error.message)
+          }
+        })
+      )
+
+      if (numberOfFilesMovedFail === items.length) {
+        toast.error('Failed to move items')
+      } else {
+        const movedFiles = items.length - numberOfFilesMovedFail
+        toast(
+          `Successfully moved ${movedFiles} item${movedFiles > 1 ? 's' : ''} to ${formattedNewPathToFile.length > 0 ? formattedNewPathToFile : 'the root of your bucket'}`
+        )
+      }
+
+      toast.dismiss(toastId)
+
+      // Clear selected items after successful move operation
+      state.clearSelectedItems()
+
+      // TODO: invalidate the file preview cache when moving files
+      await state.validateAndUpdateOpenedFolders()
+    },
+
+    // New function to handle folder moves by creating new folder and moving contents
+    moveFolderWithContents: async (folder: StorageItemWithColumn, targetPath: string) => {
+      const folderName = folder.name
+      const sourcePath = state.openedFolders
+        .slice(0, folder.columnIndex)
+        .map((folder) => folder.name)
+        .join('/')
+
+      const sourceFullPath = sourcePath.length > 0 ? `${sourcePath}/${folderName}` : folderName
+      const targetFullPath = targetPath.length > 0 ? `${targetPath}/${folderName}` : folderName
+
+      // Add folder to the "being moved" set for visual feedback
+      state.addFolderBeingMoved(targetFullPath)
+
+      // Create a placeholder file to create the new folder
+      const placeholderContent = new Blob([''], { type: 'text/plain' })
+
+      try {
+        // Check if supabaseClient is available
+        if (!state.supabaseClient) {
+          throw new Error('Supabase Client is not available')
+        }
+
+        // Check if selectedBucket and projectRef are available
+        if (!state.selectedBucket || !state.selectedBucket.id) {
+          throw new Error('Selected bucket is not available')
+        }
+
+        if (!state.projectRef) {
+          throw new Error('Project reference is not available')
+        }
+
+        // Create the new folder by uploading a placeholder file
+        const { data: uploadData, error: uploadError } = await (
+          await state.supabaseClient()
+        ).storage
+          .from(state.selectedBucket.id)
+          .upload(targetFullPath + '/.folder-placeholder', placeholderContent, {
+            upsert: false,
+          })
+
+        if (uploadError) {
+          throw new Error(`Failed to create target folder: ${uploadError.message}`)
+        }
+
+        // Recursively discover all files and folders in the source directory
+        const getAllFilesRecursively = async (
```

Comment:
this seems the same (?) as what `getAllItemsAlongFolder` is trying to do maybe? if so, could we de-dupe?